python/tests/aggregate_tests/aggtst_base.py

```diff
@@ -181,15 +180,75 @@ def run_tests(self):
         pipeline.start()
 
         for table in self.tables:
-            data = table.get_data()
-            pipeline.input_json(table.name, data, update_format="insert_delete")
+            pipeline.input_json(
+                table.name, table.get_data(), update_format="insert_delete"
+            )
 
         pipeline.wait_for_completion(shutdown=False, timeout_s=3600)
+
         for view in self.views:
             view.validate(pipeline)
 
         pipeline.shutdown()
 
+    def assert_expected_error(self, view: View, actual_exception: Exception):
+        """Validate the error produced by the failing pipeline with the expected error type"""
+        expected_substring = (
+            str(getattr(view, "expected_error_type", "") or "").strip().lower()
+        )
+        actual_message = str(actual_exception).strip().lower()
+
+        if DEBUG:
+            print(
+                f"[DEBUG] View `{view.name}` expected error substring: '{expected_substring}'"
+            )
+            print(
+                f"[DEBUG] View `{view.name}` received error message:\n{actual_message}"
+            )
+
+        if expected_substring not in actual_message:
+            raise AssertionError(
+                f"\n[FAIL] failed view: {view.name} did not produce expected error substring.\n"
+                f"Expected to find: '{expected_substring}'\n"
+                f"Received error message:\n{actual_message}"
+            )  # Validate based on: does the error received contain the expected substring?
+
+        if DEBUG:
+            print(f"[PASS] View `{view.name}` failed as expected.")
+
+    def run_expected_failures(self):
+        """Run each view that is expected to fail in a separate pipeline"""
+        original_views = self.views
+        # List of views that contain the attribute: expected error type
+        failing_views = [v for v in self.views if v.expected_error_type]
+        for view in failing_views:
+            self.views = [view]
+            sql = (
+                self.generate_sql()
```

Comment:
let's change generate_sql to take a list of views, then we don't need to save and restore self.views.
I expect that this will make the function `@static`